File: tests/pytests/spatialdb/TestSimpleGridDB.py
#!/usr/bin/env nemesis
#
# =================================================================================================
# This code is part of SpatialData, developed through the Computational Infrastructure
# for Geodynamics (https://github.com/geodynamics/spatialdata).
#
# Copyright (c) 2010-2025, University of California, Davis and the SpatialData Development Team.
# All rights reserved.
#
# See https://mit-license.org/ and LICENSE.md and for license information. 
# =================================================================================================

import unittest

import numpy
from spatialdata.testing.TestCases import make_suite
from spatialdata.geocoords.CSCart import CSCart
from spatialdata.spatialdb.SimpleGridAscii import SimpleGridAscii
from spatialdata.spatialdb.SimpleGridDB import SimpleGridDB


class TestSimpleGridDB(unittest.TestCase):

    def setUp(self):
        db = SimpleGridDB()
        db.inventory.label = "test"
        db.inventory.queryType = "nearest"
        db.inventory.filename = "data/grid.spatialdb"
        db._configure()
        self._db = db

    def test_database(self):
        locs = numpy.array([[3.5, 0.1, 0.8], [8.0, -0.4, 1.8]], numpy.float64)
        cs = CSCart()
        cs._configure()
        queryVals = ["two", "one"]
        dataE = numpy.array([[0.0220, 1.04], [0.0222, 1.02]], numpy.float64)
        errE = [0, 0]

        db = self._db
        db.open()
        db.setQueryValues(queryVals)
        data = numpy.zeros(dataE.shape, dtype=numpy.float64)
        err = []
        nlocs = locs.shape[0]
        for i in range(nlocs):
            e = db.query(data[i, :], locs[i, :], cs)
            err.append(e)
        db.close()

        self.assertEqual(len(errE), len(err))
        for vE, v in zip(errE, err):
            self.assertEqual(vE, v)

        self.assertEqual(len(dataE.shape), len(data.shape))
        for dE, d in zip(dataE.shape, data.shape):
            self.assertEqual(dE, d)
        for vE, v in zip(numpy.reshape(dataE, -1), numpy.reshape(data, -1)):
            self.assertAlmostEqual(vE, v, 6)

    def test_databasemulti(self):
        locs = numpy.array([[2.3, -0.1, 1.8], [2.0,  0.4, -2.0]], numpy.float64)
        cs = CSCart()
        cs._configure()
        queryVals = ["two", "one"]
        dataE = numpy.array([[0.0207, 1.17], [0.0209, 1.15]], numpy.float64)
        errE = numpy.array([0, 0], numpy.int32)

        db = self._db
        db.open()
        db.setQueryValues(queryVals)
        data = numpy.zeros(dataE.shape, dtype=numpy.float64)
        err = numpy.zeros(errE.shape, dtype=numpy.int32)
        db.multiquery(data, err, locs, cs)
        db.close()

        self.assertEqual(len(errE), len(err))
        for vE, v in zip(errE, err):
            self.assertEqual(vE, v)

        self.assertEqual(len(dataE.shape), len(data.shape))
        for dE, d in zip(dataE.shape, data.shape):
            self.assertEqual(dE, d)
        for vE, v in zip(numpy.reshape(dataE, -1), numpy.reshape(data, -1)):
            self.assertAlmostEqual(vE, v, 6)

    def test_io_3d(self):
        from spatialdata.spatialdb.SimpleGridAscii import createWriter

        filename = "data/gridio3d.spatialdb"
        x = numpy.array([-2.0, 0.0, 3.0], dtype=numpy.float64)
        y = numpy.array([0.0, 1.0], dtype=numpy.float64)
        z = numpy.array([-2.0, -1.0, 2.0], dtype=numpy.float64)

        points = numpy.array([
            [-2.0,  0.0, -2.0],
            [-2.0,  1.0, -2.0],
            [-2.0,  0.0, -1.0],
            [-2.0,  1.0, -1.0],
            [-2.0,  0.0,  2.0],
            [-2.0,  1.0,  2.0],
            [0.0,  0.0, -2.0],
            [0.0,  1.0, -2.0],  # query (5.7, 8.2)
            [0.0,  0.0, -1.0],
            [0.0,  1.0, -1.0],
            [0.0,  0.0,  2.0],
            [0.0,  1.0,  2.0],
            [3.0,  0.0, -2.0],
            [3.0,  1.0, -2.0],
            [3.0,  0.0, -1.0],
            [3.0,  1.0, -1.0],
            [3.0,  0.0,  2.0],
            [3.0,  1.0,  2.0],
        ], dtype=numpy.float64)
        one = numpy.array([6.6, 5.5, 2.3, 5.7, 6.3, 3.4, 7.2, 5.7, 3.4, 5.7, 9.4,
                           7.2, 4.8, 9.2, 5.8, 4.7, 7.8, 2.9], dtype=numpy.float64)
        two = numpy.array([3.4, 6.7, 4.1, 2.0, 6.7, 6.4, 6.8, 8.2, 9.8, 2.3, 8.5,
                           9.3, 7.5, 8.3, 8.5, 8.9, 6.2, 8.3], dtype=numpy.float64)

        cs = CSCart()
        cs._configure()

        writer = createWriter(filename)
        writer.write({'points': points,
                      'x': x,
                      'y': y,
                      'z': z,
                      'coordsys': cs,
                      'data_dim': 3,
                      'values': [{'name': "one",
                                  'units': "m",
                                  'data': one},
                                 {'name': "two",
                                  'units': "m",
                                  'data': two},
                                 ]})

        db = SimpleGridDB()
        db.inventory.label = "test"
        db.inventory.queryType = "nearest"
        db.inventory.filename = filename
        db._configure()
        self._db = db

        locs = numpy.array([[0.1, 0.95, -1.8]], numpy.float64)
        cs = CSCart()
        cs._configure()
        queryVals = ["two", "one"]
        dataE = numpy.array([[8.2, 5.7]], numpy.float64)
        errE = [0]

        db = self._db
        db.open()
        db.setQueryValues(queryVals)
        data = numpy.zeros(dataE.shape, dtype=numpy.float64)
        err = []
        nlocs = locs.shape[0]
        for i in range(nlocs):
            e = db.query(data[i, :], locs[i, :], cs)
            err.append(e)
        db.close()

        self.assertEqual(len(errE), len(err))
        for vE, v in zip(errE, err):
            self.assertEqual(vE, v)

        self.assertEqual(len(dataE.shape), len(data.shape))
        for dE, d in zip(dataE.shape, data.shape):
            self.assertEqual(dE, d)
        for vE, v in zip(numpy.reshape(dataE, -1), numpy.reshape(data, -1)):
            self.assertAlmostEqual(vE, v, 6)

    def test_io_2d(self):
        from spatialdata.spatialdb.SimpleGridAscii import createWriter

        filename = "data/gridio2d.spatialdb"
        x = numpy.array([-2.0, 0.0, 3.0], dtype=numpy.float64)
        y = numpy.array([0.0, 1.0], dtype=numpy.float64)

        points = numpy.array([
            [-2.0,  0.0],
            [-2.0,  1.0],
            [0.0,  0.0],
            [0.0,  1.0],  # query (5.7, 8.2)
            [3.0,  0.0],
            [3.0,  1.0],
        ], dtype=numpy.float64)
        one = numpy.array([6.6, 5.5, 7.2, 5.7, 4.8, 9.2], dtype=numpy.float64)
        two = numpy.array([3.4, 6.7, 6.8, 8.2, 7.5, 8.3], dtype=numpy.float64)

        cs = CSCart()
        cs.inventory.spaceDim = 2
        cs._configure()

        writer = createWriter(filename)
        writer.write({'points': points,
                      'x': x,
                      'y': y,
                      'coordsys': cs,
                      'data_dim': 2,
                      'values': [{'name': "one",
                                  'units': "m",
                                  'data': one},
                                 {'name': "two",
                                  'units': "m",
                                  'data': two},
                                 ]})

        db = SimpleGridDB()
        db.inventory.label = "test"
        db.inventory.queryType = "nearest"
        db.inventory.filename = filename
        db._configure()
        self._db = db

        locs = numpy.array([[0.1, 0.95]], numpy.float64)
        queryVals = ["two", "one"]
        dataE = numpy.array([[8.2, 5.7]], numpy.float64)
        errE = [0]

        db = self._db
        db.open()
        db.setQueryValues(queryVals)
        data = numpy.zeros(dataE.shape, dtype=numpy.float64)
        err = []
        nlocs = locs.shape[0]
        for i in range(nlocs):
            e = db.query(data[i, :], locs[i, :], cs)
            err.append(e)
        db.close()

        self.assertEqual(len(errE), len(err))
        for vE, v in zip(errE, err):
            self.assertEqual(vE, v)

        self.assertEqual(len(dataE.shape), len(data.shape))
        for dE, d in zip(dataE.shape, data.shape):
            self.assertEqual(dE, d)
        for vE, v in zip(numpy.reshape(dataE, -1), numpy.reshape(data, -1)):
            self.assertAlmostEqual(vE, v, 6)


def load_tests(loader, tests, pattern):
    TEST_CLASSES = [TestSimpleGridDB]
    return make_suite(TEST_CLASSES, loader)


if __name__ == "__main__":
    unittest.main(verbosity=2)


# End of file

File: tests/libtests/spatialdb/TestSimpleGridDB.cc
// =================================================================================================
// This code is part of SpatialData, developed through the Computational Infrastructure
// for Geodynamics (https://github.com/geodynamics/spatialdata).
//
// Copyright (c) 2010-2024, University of California, Davis and the SpatialData Development Team.
// All rights reserved.
//
// See https://mit-license.org/ and LICENSE.md and for license information.
// =================================================================================================

#include <portinfo>

#include "TestSimpleGridDB.hh" // Implementation of class methods

#include "spatialdata/spatialdb/SimpleGridDB.hh" // USES SimpleGridDB
#include "spatialdata/spatialdb/SimpleGridAscii.hh" // USES SimpleGridAscii

#include "spatialdata/geocoords/CSCart.hh" // USE CSCart

#include "catch2/catch_test_macros.hpp"
#include "catch2/matchers/catch_matchers_floating_point.hpp"

#include <cmath> // USES fabs()

// ----------------------------------------------------------------------
// Constructor.
spatialdata::spatialdb::TestSimpleGridDB::TestSimpleGridDB(TestSimpleGridDB_Data* data) :
    _data(data) {
    assert(_data);
} // setUp


// ----------------------------------------------------------------------
// Tear down testing data.
spatialdata::spatialdb::TestSimpleGridDB::~TestSimpleGridDB(void) {
    delete _data;_data = NULL;
} // tearDown


// ----------------------------------------------------------------------
// Test constructor
void
spatialdata::spatialdb::TestSimpleGridDB::testConstructor(void) {
    SimpleGridDB db;
} // testConstructor


// ----------------------------------------------------------------------
// Test setDescription()
void
spatialdata::spatialdb::TestSimpleGridDB::testAccessors(void) {
    SimpleGridDB db;
    const std::string description("database 2");
    const std::string filename("mydb.spatialdb");

    db.setDescription(description.c_str());
    CHECK(description == std::string(db.getDescription()));

    db.setFilename(filename.c_str());
    CHECK(filename == db._filename);

    // Test query type.
    CHECK(SimpleGridDB::NEAREST == db._queryType);

    db.setQueryType(SimpleGridDB::LINEAR);
    CHECK(SimpleGridDB::LINEAR == db._queryType);

    db.setQueryType(SimpleGridDB::NEAREST);
    CHECK(SimpleGridDB::NEAREST == db._queryType);
} // testAccessors


// ----------------------------------------------------------------------
// Test _search()
void
spatialdata::spatialdb::TestSimpleGridDB::testSearch(void) {
    const size_t numX = 5;
    const double x[numX] = {
        -3.0, -1.0, 0.0, 5.0, 8.0,
    };
    const double tolerance = 1.0e-6;

    SimpleGridDB db;

    db.setQueryType(SimpleGridDB::NEAREST);

    // Test data and expected results
    const double xA = -20.0;
    const double indexA = 0.0;
    double index = db._search(xA, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexA, tolerance));

    const double xB = -1.0;
    const double indexB = 1.0;
    index = db._search(xB, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexB, tolerance));

    const double xC = +3.0;
    const double indexC = 2.6;
    index = db._search(xC, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexC, tolerance));

    const double xD = -20.0;
    const double indexD = 0.0;
    index = db._search(xD, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexD, tolerance));

    db.setQueryType(SimpleGridDB::LINEAR);

    // Test data and expected results
    const double xE = -20.0;
    const double indexE = -1.0;
    index = db._search(xE, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexE, tolerance));

    const double xF = -1.0;
    const double indexF = 1.0;
    index = db._search(xF, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexF, tolerance));

    const double xG = +3.0;
    const double indexG = 2.6;
    index = db._search(xG, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexG, tolerance));

    const double xH = -20.0;
    const double indexH = -1.0;
    index = db._search(xH, x, numX);
    CHECK_THAT(index, Catch::Matchers::WithinAbs(indexH, tolerance));
} // testSearch


// ----------------------------------------------------------------------
// Test _getDataIndex()
void
spatialdata::spatialdb::TestSimpleGridDB::testDataIndex(void) {
    SimpleGridDB db;
    db._numX = 4;
    db._numY = 3;
    db._numZ = 5;
    db._numValues = 10;

    CHECK(size_t(0) == db._getDataIndex(0, db._numX, 0, db._numY, 0, db._numZ));
    CHECK(size_t(1*3*4*10) == db._getDataIndex(0, db._numX, 0, db._numY, 1, db._numZ));

    CHECK(size_t(1*4*10) == db._getDataIndex(0, db._numX, 1, db._numY, 0, db._numZ));
    CHECK(size_t(4*3*4*10 + 1*4*10) == db._getDataIndex(0, db._numX, 1, db._numY, 4, db._numZ));

    CHECK(size_t(1*10) == db._getDataIndex(1, db._numX, 0, db._numY, 0, db._numZ));
    CHECK(size_t(3*4*3*10 + 1*4*10 + 2*10) == db._getDataIndex(2, db._numX, 1, db._numY, 3, db._numZ));
} // testDataIndex


// ----------------------------------------------------------------------
// Test getNamesDBValues().
void
spatialdata::spatialdb::TestSimpleGridDB::testGetNamesDBValues(void) {
    assert(_data);

    SimpleGridDB db;
    _setupDB(&db);

    const char** valueNames = NULL;
    size_t numValues = 0;
    db.getNamesDBValues(&valueNames, &numValues);
    REQUIRE(_data->numValues == numValues);

    for (size_t i = 0; i < numValues; ++i) {
        CHECK(std::string(_data->names[i]) == std::string(valueNames[i]));
    } // for
    delete[] valueNames;valueNames = NULL;
    numValues = 0;
} // testGetDBValues


// ----------------------------------------------------------------------
// Test query() using nearest neighbor.
void
spatialdata::spatialdb::TestSimpleGridDB::testQueryNearest(void) {
    assert(_data);

    SimpleGridDB db;
    _setupDB(&db);
    db.setQueryType(SimpleGridDB::NEAREST);
    _checkQuery(db, _data->names, _data->queryNearest, 0, _data->numQueries, _data->spaceDim, _data->numValues);
} // _testQueryNearest


// ----------------------------------------------------------------------
// Test query() using linear interpolation.
void
spatialdata::spatialdb::TestSimpleGridDB::testQueryLinear(void) {
    assert(_data);

    SimpleGridDB db;
    _setupDB(&db);
    db.setQueryType(SimpleGridDB::LINEAR);
    _checkQuery(db, _data->names, _data->queryLinear, _data->errFlags, _data->numQueries, _data->spaceDim, _data->numValues);
} // _testQueryLinear


// ----------------------------------------------------------------------
// Test read().
void
spatialdata::spatialdb::TestSimpleGridDB::testRead(void) {
    assert(_data);

    SimpleGridDB db;
    db.setFilename(_data->filename);
    db.open();

    CHECK(_data->spaceDim == db._spaceDim);
    CHECK(_data->dataDim == db._dataDim);
    REQUIRE(_data->numValues == db._numValues);

    const size_t numValues = _data->numValues;
    for (size_t i = 0; i < numValues; ++i) {
        CHECK(std::string(_data->names[i]) == db._names[i]);
        CHECK(std::string(_data->units[i]) == db._units[i]);
    } // for

    CHECK(_data->numX == db._numX);
    CHECK(_data->numY == db._numY);
    CHECK(_data->numZ == db._numZ);

    const double tolerance = 1.0e-6;
    size_t totalSize = 1;
    if (_data->numX > 1) {
        totalSize *= _data->numX;
        for (size_t i = 0; i < _data->numX; ++i) {
            CHECK_THAT(db._x[i], Catch::Matchers::WithinAbs(_data->dbX[i], tolerance));
        } // for
    } // if
    if (_data->numY > 1) {
        totalSize *= _data->numY;
        for (size_t i = 0; i < _data->numY; ++i) {
            CHECK_THAT(db._y[i], Catch::Matchers::WithinAbs(_data->dbY[i], tolerance));
        } // for
    } // if
    if (_data->numZ > 1) {
        totalSize *= _data->numZ;
        for (size_t i = 0; i < _data->numZ; ++i) {
            CHECK_THAT(db._z[i], Catch::Matchers::WithinAbs(_data->dbZ[i], tolerance));
        } // for
    } // if

    for (size_t i = 0; i < totalSize; ++i) {
        CHECK_THAT(db._data[i], Catch::Matchers::WithinAbs(_data->dbData[i], tolerance));
    } // for

    CHECK(_data->numValues == db._querySize);
    for (size_t i = 0; i < _data->numValues; ++i) {
        CHECK(i == db._queryValues[i]);
    } // for

    db.close();
} // testRead


// ----------------------------------------------------------------------
// Populate database with data.
void
spatialdata::spatialdb::TestSimpleGridDB::_setupDB(SimpleGridDB* const db) {
    assert(db);
    assert(_data);
    assert(_data->numValues > 0);

    delete[] db->_x;db->_x = NULL;
    delete[] db->_y;db->_y = NULL;
    delete[] db->_z;db->_z = NULL;
    delete[] db->_data;db->_data = NULL;
    delete[] db->_names;db->_names = NULL;
    delete[] db->_units;db->_units = NULL;

    db->setDescription("Test database");
    db->_numValues = _data->numValues;
    db->_spaceDim = _data->spaceDim;
    db->_dataDim = _data->dataDim;
    db->_numX = _data->numX;
    db->_numY = _data->numY;
    db->_numZ = _data->numZ;

    size_t numLocs = 1;
    if (_data->numX > 1) {
        numLocs *= _data->numX;
        db->_x = new double[_data->numX];
        for (size_t i = 0; i < _data->numX; ++i) {
            db->_x[i] = _data->dbX[i];
        } // for
    } // if
    if (_data->numY > 1) {
        numLocs *= _data->numY;
        db->_y = new double[_data->numY];
        for (size_t i = 0; i < _data->numY; ++i) {
            db->_y[i] = _data->dbY[i];
        } // for
    } // if
    if (_data->numZ > 1) {
        numLocs *= _data->numZ;
        db->_z = new double[_data->numZ];
        for (size_t i = 0; i < _data->numZ; ++i) {
            db->_z[i] = _data->dbZ[i];
        } // for
    } // if

    db->_data = (numLocs > 0) ? new double[numLocs*_data->numValues] : NULL;
    for (size_t i = 0; i < numLocs*_data->numValues; ++i) {
        db->_data[i] = _data->dbData[i];
    } // for

    db->_names = (_data->numValues > 0) ? new std::string[_data->numValues] : NULL;
    for (size_t i = 0; i < _data->numValues; ++i) {
        db->_names[i] = _data->names[i];
    } // for

    db->_units = (_data->numValues > 0) ? new std::string[_data->numValues] : NULL;
    for (size_t i = 0; i < _data->numValues; ++i) {
        db->_units[i] = _data->units[i];
    } // for

    delete db->_cs;
    db->_cs = new spatialdata::geocoords::CSCart();
    db->_cs->setSpaceDim(_data->spaceDim);
} // _setupDB


// ----------------------------------------------------------------------
// Test query method by doing query and checking values returned.
void
spatialdata::spatialdb::TestSimpleGridDB::_checkQuery(SimpleGridDB& db,
                                                      const char** const names,
                                                      const double* queryData,
                                                      const int* flagsE,
                                                      const size_t numQueries,
                                                      const size_t spaceDim,
                                                      const size_t numValues) {
    assert(names);
    assert(queryData);
    assert(numQueries);
    assert(spaceDim);
    assert(numValues);

    // reverse order of vals in queries
    const char* valNames[numValues];
    for (size_t i = 0; i < numValues; ++i) {
        valNames[numValues-i-1] = names[i];
    }
    db.setQueryValues(valNames, numValues);

    double* vals = (0 < numValues) ? new double[numValues] : NULL;
    const double tolerance = 1.0e-06;

    const size_t locSize = spaceDim + numValues;
    spatialdata::geocoords::CSCart csCart;
    csCart.setSpaceDim(spaceDim);
    for (size_t iQuery = 0; iQuery < numQueries; ++iQuery) {
        const double* coords = &queryData[iQuery*locSize];
        const double* valsE = &queryData[iQuery*locSize+spaceDim];
        const int err = db.query(vals, numValues, coords, spaceDim, &csCart);
        if (0 != flagsE) {
            REQUIRE(flagsE[iQuery] == err);
        } else {
            REQUIRE(!err);
            for (size_t iVal = 0; iVal < numValues; ++iVal) {
                const double valueE = valsE[numValues-iVal-1];
                const double toleranceV = fabs(valueE) > 0.0 ? fabs(valueE) * tolerance : tolerance;
                CHECK_THAT(vals[iVal], Catch::Matchers::WithinAbs(valueE, toleranceV));
            } // for
        } // if/else
    } // for
    delete[] vals;vals = NULL;
} // _checkQuery


// ----------------------------------------------------------------------
spatialdata::spatialdb::TestSimpleGridDB_Data::TestSimpleGridDB_Data(void) :
    numX(0),
    numY(0),
    numZ(0),
    spaceDim(0),
    numValues(0),
    dataDim(0),
    dbX(NULL),
    dbY(NULL),
    dbZ(NULL),
    dbData(NULL),
    names(NULL),
    units(NULL),
    numQueries(0),
    queryNearest(NULL),
    queryLinear(NULL),
    errFlags(NULL),
    filename(NULL) {}


// ----------------------------------------------------------------------
spatialdata::spatialdb::TestSimpleGridDB_Data::~TestSimpleGridDB_Data(void) {
    numX = 0;
    numY = 0;
    numZ = 0;
    spaceDim = 0;
    numValues = 0;
    dataDim = 0;
    dbX = NULL;
    dbY = NULL;
    dbZ = NULL;
    dbData = NULL;
    names = NULL;
    units = NULL;
    numQueries = 0;
    queryNearest = NULL;
    queryLinear = NULL;
    errFlags = NULL;
    filename = NULL;
} // destructor


// End of file
